cluster-autoscaler/cloudprovider/clusterapi/internal/snapshot.go

```diff
@@ -0,0 +1,147 @@
+package internal
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	v1alpha1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	v1alpha1apis "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+type MachineSetID string
+
+type clusterSnapshot struct {
+	NodeMap             map[MachineSetID]map[string]string
+	MachineSetMap       map[MachineSetID]*clusterMachineSet
+	NodeToMachineSetMap map[string]MachineSetID
+	MachineSetNodeMap   map[MachineSetID][]string
+}
+
+func machineSetID(m *v1alpha1.MachineSet) MachineSetID {
+	return MachineSetID(fmt.Sprintf("%s/%s", m.Namespace, m.Name))
+}
+
+func getMachinesInMachineSet(m *clusterManager, ms *v1alpha1apis.MachineSet) ([]*v1alpha1apis.Machine, error) {
+	machines, err := m.clusterapi.Machines(ms.Namespace).List(v1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(ms.Spec.Selector.MatchLabels).String(),
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to list machines in namespace %s: %v", ms.Namespace, err)
+	}
+
+	names := make([]string, len(machines.Items))
+	result := make([]*v1alpha1apis.Machine, len(machines.Items))
+
+	for i := range machines.Items {
+		names[i] = machines.Items[i].Name
+		result[i] = &machines.Items[i]
+	}
+
+	glog.Infof("%d machines in machineset %s/%s: %#v", len(result), ms.Namespace, ms.Name, names)
+
+	return result, nil
+}
+
+func getMachineSetsInNamespace(m *clusterManager, namespace string) ([]*v1alpha1apis.MachineSet, error) {
+	machineSets, err := m.clusterapi.MachineSets(namespace).List(v1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to list machinesets in namespace %q: %v", namespace, err)
+	}
+
+	names := make([]string, len(machineSets.Items))
+	result := make([]*v1alpha1apis.MachineSet, len(machineSets.Items))
+
+	for i := range machineSets.Items {
+		names[i] = machineSets.Items[i].Name
+		result[i] = &machineSets.Items[i]
+	}
+
+	return result, nil
+}
+
+func getNamespaces(m *clusterManager) ([]string, error) {
+	namespaces, err := m.kubeclient.CoreV1().Namespaces().List(v1.ListOptions{})
```

Comment:
Should really be an informer.  If you are going to be managing caches use informers and listers everywhere.

Reply:
Currently work in progress to move this to informers.